src/main/kotlin/com/team4099/robot2020/subsystems/Feeder.kt

```diff
@@ -0,0 +1,152 @@
+    )
+
+    private val stopperTalon = CTREMotorControllerFactory.createDefaultTalonFX(Constants.Feeder.FEEDER_OUT_ID)
+
+    private val inBeamBreak = DigitalInput(Constants.BeamBreak.IN_BEAM_BREAK_PORT)
+
+    private var inBeamBroken = false
+        get() = inBeamBreak.get()
+
+    private var inBeamBrokenTimestamp = 0.0
+
+    private val outBeamBreak = DigitalInput(Constants.BeamBreak.OUT_BEAM_BREAK_PORT)
+
+    private var outBeamBroken = false
+        get() = outBeamBreak.get()
+
+    private var outBeamBrokenTimestamp = 0.0
+
+    // take this out after adding one ballCount in superstructure to work with intake
+    var ballCount = 0
+
+    var feederState = FeederState.IDLE
+        set(value) {
+            if (value != field) {
+                HelixEvents.addEvent("FEEDER", "state changed to $value")
+            }
+        }
+    private var feederInPower = 0.0
+        set(value) {
+            if (value != field) {
+                inMasterSparkMax.set(value)
+            }
+            field = value
+        }
+    private var feederStopperPower = 0.0
+        set(value) {
+            if (value != field) {
+                stopperTalon.set(ControlMode.PercentOutput, value)
+            }
+            field = value
+        }
+
+    init {
+        inMasterSparkMax.inverted = false
+        inSlaveSparkMax.inverted = true
+
+        stopperTalon.inverted = true
+    }
+
+    enum class FeederState {
+        HOLD, INTAKE, SHOOT, EXHAUST, IDLE
+    }
+
+    override fun outputTelemetry() {
+        SmartDashboard.putString("feeder/feederState", feederState.toString())
+        SmartDashboard.putNumber("feeder/feederInPower", feederInPower)
+        SmartDashboard.putNumber("feeder/feederStopperPower", feederStopperPower)
+    }
+
+    override fun checkSystem() {}
+
+    override fun registerLogging() {
+        HelixLogger.addSource("Feeder Master Motor Power") { inMasterSparkMax.appliedOutput }
+        HelixLogger.addSource("Feeder Slave Motor Power") { inSlaveSparkMax.appliedOutput }
+        HelixLogger.addSource("Feeder Stopper Motor Power") { stopperTalon.motorOutputPercent }
+
+        HelixLogger.addSource("Feeder Master Motor Current") { inMasterSparkMax.outputCurrent }
+        HelixLogger.addSource("Feeder Slave Motor Current") { inSlaveSparkMax.outputCurrent }
+        HelixLogger.addSource("Feeder Stopper Motor current") { stopperTalon.supplyCurrent }
+
+        HelixLogger.addSource("Feeder State") { feederState }
+    }
+
+    override fun zeroSensors() { }
+
+    override fun onStart(timestamp: Double) {
+        feederState = FeederState.HOLD
+    }
+
+    @Synchronized
+    override fun onLoop(timestamp: Double, dT: Double) {
+
+        if (inBeamBroken) {
+            if (inBeamBrokenTimestamp == -1.0) {
+                inBeamBrokenTimestamp = timestamp
+            }
+        } else {
+            if (timestamp - inBeamBrokenTimestamp >= Constants.BeamBreak.IN_BEAM_BROKEN_BALL_TIME) {
```

Comment:
seems like you're only checking to increment ball count when the beam stops being broken -- perhaps I'm not understanding right, but isn't the point of having that time threshold so that if two balls come in in succession (beam is broken the whole time) we account for two? If so I think just moving this if statement out of the else will fix it

Reply:
or actually you could also do integer division `IN_BEAM_BROKEN_BALL_TIME / (timestamp - inBeamBrokenTimestamp)` as the number of balls that just passed through? idk